src/ext/engine_hooks.h

```diff
@@ -165,4 +173,36 @@ void ddtrace_close_all_open_spans(TSRMLS_D);
 void ddtrace_close_all_open_spans(void);
 #endif
 
+#if PHP_VERSION_ID >= 80000
+inline zend_class_entry *ddtrace_get_exception_base(zval *object) {
+    return (Z_OBJCE_P(object) == zend_ce_exception || instanceof_function_slow(Z_OBJCE_P(object), zend_ce_exception))
+               ? zend_ce_exception
+               : zend_ce_error;
+}
+#define GET_PROPERTY(object, id) \
+    zend_read_property_ex(ddtrace_get_exception_base(object), Z_OBJ_P(object), ZSTR_KNOWN(id), 1, &rv)
+#elif PHP_VERSION_ID >= 70000
+inline zend_class_entry *ddtrace_get_exception_base(zval *object) {
+    return instanceof_function(Z_OBJCE_P(object), zend_ce_exception) ? zend_ce_exception : zend_ce_error;
+}
+#if PHP_VERSION_ID < 70100
+#define ZEND_STR_MESSAGE "message"
+#define GET_PROPERTY(object, name) \
+    zend_read_property(ddtrace_get_exception_base(object), (object), name, sizeof(name) - 1, 1, &rv)
+#elif PHP_VERSION_ID < 70200
+#define GET_PROPERTY(object, id) \
+    zend_read_property_ex(ddtrace_get_exception_base(object), (object), CG(known_strings)[id], 1, &rv)
+#else
+#define GET_PROPERTY(object, id) \
+    zend_read_property_ex(ddtrace_get_exception_base(object), (object), ZSTR_KNOWN(id), 1, &rv)
+#endif
+#endif
+
+#if PHP_VERSION_ID < 70000
+inline zval *ddtrace_exception_get_entry(zval *object, char *name, int name_len TSRMLS_DC) {
+    zend_class_entry *exception_ce = zend_exception_get_default(TSRMLS_C);
+    return zend_read_property(exception_ce, object, name, name_len, 1 TSRMLS_CC);
+}
+#endif
```

Comment:
This is... less than ideal. 😖 🐢 But the idea is that we will split `engine_hooks.h` for each PHP version in the next PR.